include/crab/domains/mru_region_domain.hpp

```diff
@@ -0,0 +1,2270 @@
+      ghost_variables_t::mk_renaming_map(m_rev_var_map, get_type_fn,
+                                         renaming_map);
+      m_alloc.add_renaming_map(renaming_map);
+      o << "Mem=\t" << m_mem << ",\n"
+        << "( CacheLine=\t" << m_cache_lines << ",\n"
+        << "Regs=\t" << m_regs << ",\n"
+        << "Addrs=\t" << m_addrs << " )\n";
+      m_alloc.clear_renaming_map();
+    }
+  }
+  /**------------- End domain API definitions -------------------**/
+
+  /**------------- Begin MRU API definitions -------------------**/
+
+  // The expand performs a copy of region vars from mem into cache_lines
+  //  i.e. by given a most recently used object o, we copy all regions with o
+  //  in mem into the cache:
+  // For instance, o used V1, V2:
+  //    mem: { 1 <= V1 <= 2; V1 < V2; ... }
+  // The cache line will be:
+  //    cache_lines: { 1 <= V1 <= 2; V1 < V2; }
+  void expand(const variable_t &rgn) {
+    variable_vector_t Vs_for_new_obj;
+    auto it_var = m_rev_obj_rgn_map.find(rgn);
+    if (it_var == m_rev_obj_rgn_map.end()) {
+      // this should not happen
+      CRAB_ERROR("mru_region_domain::expand the mru object should be found");
+    }
+    variable_t representative = it_var->second;
+    auto it_rev_vars = m_obj_rgn_map.find(representative);
+    if (it_rev_vars == m_obj_rgn_map.end()) {
+      // this should not happen
+      CRAB_ERROR("mru_region_domain::expand the mru object should be found");
+    }
+    Vs_for_new_obj = it_rev_vars->second;
+
+    // expand does not need to perform base variable renaming
+    mru_region_domain_t tmp = mru_region_domain_t(*this);
+    tmp.project(Vs_for_new_obj);
+
+    // perform expand
+    m_cache_lines |= tmp.m_mem;
+
+    CRAB_LOG("mru-region", crab::outs() << *this << "\n");
+  }
+
+  // The cache is used by given an allocated object.
+  //  The fold operation performs smash all regions (fields) into the summary
+  //  object in mem. That is, for each region used in cache line, join with
+  //  region in mem.
+  //  In addition, we also update regs used in cache via reduction.
+  void fold() {
+    variable_vector_t Vs;
+    region_vars_in_mem(Vs);
+
+    // Put together cache lines and regs
+    // Step 1.
+    //  Perform regs: { r == Cj ... }
+    //  constrain with cache_lines: { Ci <= Cj <= Ck ... }, i != j != k
+    // the reduced cache_lines will be:
+    //    { Ci <= Cj <= Ck; Ci <= r <= Ck; ...  }
+    base_abstract_domain_t reduced = cache_lines_regs_reduce();
+
+    // Step 2.
+    //  Our design is joining cache line with memory.
+    //  We need the projection before join because we only fold regions & regs
+    //  We do not affect other properties in mem.
+    variable_vector_t vars_cache;
+    vars_in_cache(vars_cache); // vars for cache lines + registers
+    mru_region_domain_t A = mru_region_domain_t(*this);
+    A.project(vars_cache); // memory object represented in the cache
+    A.m_mem |= reduced;    // join cache with mem,
```

Comment:
I think before join we need to rename `Ci` (cache names) into `Ri` (mem names).
Update: I also wrote something in `expand`. Is your intention that cache and mem have the same region variable names?

Reply:
Also, `vars_cache` seems to contain `Ci` variables so if we project `A` (copy of `m_mem`) on those variables you will probably get top since `A` does not contain `Ci` variables.